diff --git a/node_modules/@azure/static-web-apps-cli/dist/msha/handlers/function.handler.js b/node_modules/@azure/static-web-apps-cli/dist/msha/handlers/function.handler.js
index 2cc7edb..3d7c701 100644
--- a/node_modules/@azure/static-web-apps-cli/dist/msha/handlers/function.handler.js
+++ b/node_modules/@azure/static-web-apps-cli/dist/msha/handlers/function.handler.js
@@ -19,8 +19,10 @@ core_1.registerProcessExit(() => {
 });
 function injectHeaders(req, host) {
     core_1.logger.silly(`injecting headers to Functions request:`);
-    req.setHeader("x-ms-original-url", encodeURI(new URL(req.path, host).toString()));
-    core_1.logger.silly(` - x-ms-original-url: ${chalk_1.default.yellow(req.getHeader("x-ms-original-url"))}`);
+    if (!req.getHeader("x-ms-original-url")) {
+        req.setHeader("x-ms-original-url", encodeURI(new URL(req.path, host).toString()));
+        core_1.logger.silly(` - x-ms-original-url: ${chalk_1.default.yellow(req.getHeader("x-ms-original-url"))}`);
+    }
     // generate a fake correlation ID
     req.setHeader("x-ms-request-id", `SWA-CLI-${Math.random().toString(36).substring(2).toUpperCase()}`);
     core_1.logger.silly(` - x-ms-request-id: ${chalk_1.default.yellow(req.getHeader("x-ms-request-id"))}`);
diff --git a/node_modules/@azure/static-web-apps-cli/dist/msha/middlewares/response.middleware.js b/node_modules/@azure/static-web-apps-cli/dist/msha/middlewares/response.middleware.js
index 5ccfe06..0875b3f 100644
--- a/node_modules/@azure/static-web-apps-cli/dist/msha/middlewares/response.middleware.js
+++ b/node_modules/@azure/static-web-apps-cli/dist/msha/middlewares/response.middleware.js
@@ -21,6 +21,7 @@ function getResponse(req, res, matchedRoute, userConfig, isFunctionRequest) {
         return routes_engine_1.applyRedirectResponse(req, res, matchedRoute);
     }
     if (rewrite) {
+        req.headers["x-ms-original-url"] = encodeURI(req.url);
         req.url = rewrite;
     }
     if ([403, 401].includes(statusCodeToServe)) {
